### DISCUSSION TITLE
lc/54/

### DISCUSSION THREAD

**Discussion by giscus** [doocs/leetcode#2519] [Announcements]
# lc/54/

多种编程语言实现 LeetCode、《剑指 Offer（第 2 版）》、《程序员面试金典（第 6 版）》题解

https://leetcode.doocs.org/lc/54/

**Comment by hopeckel** [comment]
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        if (matrix == null) {
            return null;
        }

        List<Integer> res = new ArrayList<>();
        int m = matrix.length;
        int n = matrix[0].length;
        int eleNum = 1; //元素个数从1开始计数
        int top = 0;
        int bottom = m - 1;
        int left = 0;
        int right = n - 1;
        while (eleNum <= m * n) {
            for (int i = left; i <= right && eleNum <= m * n; i++) {
                res.add(matrix[top][i]);
                eleNum++;
            }
            top++;
            for (int i = top; i <= bottom && eleNum <= m * n; i++) {
                res.add(matrix[i][right]);
                eleNum++;
            }
            right--;
            for (int i = right; i >= left && eleNum <= m * n; i--) {
                res.add(matrix[bottom][i]);
                eleNum++;
            }
            bottom--;
            for (int i = bottom; i >= top && eleNum <= m * n; i--) {
                res.add(matrix[i][left]);
                eleNum++;
            }
            left++;
        }

        return res;
    }
}
